### PR TITLE
Closes #46 | Create dataloader for Flores200

### LINKED ISSUE #46: Create dataset loader for FLORES-200
Dataloader name: `flores200/flores200.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?flores200

|     Dataset| flores200 |
|-------------|---|
| Description | The creation of FLORES-200 doubles the existing language coverage of FLORES-101. Given the nature of the new languages, which have less standardization and require more specialized professional translations, the verification process became more complex. This required modifications to the translation workflow. FLORES-200 has several languages which were not translated from English. Specifically, several languages were translated from Spanish, French, Russian and Modern Standard Arabic. Moreover, FLORES-200 also includes two script alternatives for four languages. |
| Subsets | - |
| Languages | ace, ban, bjn, bug, ceb, ilo, ind, jav, kac, khm, lao, lus, min, mya, pag, shn, sun, tam, tgl, tha, vie, war, zsm, yue |
| Tasks | Machine Translation |
| License | Creative Commons Attribution Non Commercial 4.0 (cc-by-nc-4.0) |
| Homepage | https://github.com/facebookresearch/flores/blob/main/flores200/README.md |
| HF URL | - |
| Paper URL | https://arxiv.org/abs/2106.03193 |

### PR DESCRIPTION
**Pull request by Enliven26**
Closes [#46](https://github.com/SEACrowd/seacrowd-datahub/issues/46)

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+            first_lang_name = f"{first_lang}_{script_names}"
+
+            for second_lang in _LANGUAGES:
+
+                if first_lang == second_lang:
+                    continue
+
+                second_lang_name = f"{second_lang}_{script_names}"
+
+                subset_id = f"{first_lang}_{second_lang}_{script_names}"
```
Comment (holylovenia) [inline seacrowd/sea_datasets/flores200/flores200.py:359]: Hi @Enliven26, may I know why you didn't use `subset_id = f"{first_lang_name}_{second_lang_name}"`?

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+
+                BUILDER_CONFIGS.append(
+                    Flores200SeacrowdConfig(
+                        name=f"{subset_id}_source",
```
Comment (holylovenia) [inline seacrowd/sea_datasets/flores200/flores200.py:363]: Please prepend the `{_DATASETNAME}`.

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+                    )
+                )
+
+                seacrowd_schema_config: list[SEACrowdConfig] = []
+
+                for seacrowd_schema in _SUPPORTED_SCHEMA_STRINGS:
+
+                    seacrowd_schema_config.append(
+                        Flores200SeacrowdConfig(
+                            name=f"{subset_id}_{seacrowd_schema}",
```
Comment (holylovenia) [inline seacrowd/sea_datasets/flores200/flores200.py:380]: Please prepend the `{_DATASETNAME}`.

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+                            subset_id=subset_id,
+                            first_language_name=first_lang_name,
+                            second_language_name=second_lang_name,
+                            script_language=script_names,
+                        )
+                    )
+
+                BUILDER_CONFIGS.extend(seacrowd_schema_config)
+
+    DEFAULT_CONFIG_NAME = f"{_LANGUAGES[0]}{_LANGUAGES[1]}_{_SCRIPT_LANGUAGE_NAMES[0]}_source"
```
Comment (holylovenia) [inline seacrowd/sea_datasets/flores200/flores200.py:393]: Please prepend the `{_DATASETNAME}`.

**Review by holylovenia — CHANGES_REQUESTED** [review]
Hi @Enliven26, thanks for your contribution. I happened to pass by this dataloader and added some comments:

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 5]
```diff
@@ -0,0 +1,487 @@
...
+                    version=SEACROWD_VERSION,
+                    description=f"{_DATASETNAME} {seacrowd_schema} schema",
+                    schema=f"{seacrowd_schema}",
+                    subset_id=subset_id,
+                    first_language_name=first_lang_name,
+                    second_language_name=second_lang_name,
+                )
+            )
+
+        BUILDER_CONFIGS.extend(seacrowd_schema_config)
```
Comment (SamuelCahyawijaya) [inline seacrowd/sea_datasets/flores200/flores200.py:380]: We need to take all the language pairs where one of the language is in the `_LANGUAGES` list

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+                    "text_2": sentences[self.config.second_language_name][id_],
+                    "text_1_name": self.config.first_language_name,
+                    "text_2_name": self.config.second_language_name,
+                }
+
+        else:
+            raise ValueError(f"Invalid config: {self.config.name}")
+
+
+# This allows you to run your dataloader with `python [dataset_name].py` during development
```
Comment (SamuelCahyawijaya) [inline seacrowd/sea_datasets/flores200/flores200.py:484]: I think we can remove this part in the PR

**Review by SamuelCahyawijaya — CHANGES_REQUESTED** [review]
Hi @Enliven26 , thank you for the contribution! I think in the current scenario, you only take the language pairs with` Eng_Latn` as the source language. 

For MT datasets, we take all language pairs where at least one of the language is in the `_LANGUAGES` list. Could you be able to update it accordingly.

**Comment by SamuelCahyawijaya** [comment]
Ups @Enliven26, sorry for the last comment, I think we have a conflicting argument here. Let me discuss this internally first with @holylovenia  and the others regarding which subsets should we take 🙏🏻

**Comment by holylovenia** [comment]
> Ups @Enliven26, sorry for the last comment, I think we have a conflicting argument here. Let me discuss this internally first with @holylovenia and the others regarding which subsets should we take 🙏🏻

Hi @Enliven26, apologies for the previous contradictory advice. 🙏 After discussing, this is what we agreed on.

For parallel MT dataloaders, we agreed upon having a subset for every possible direction with at least 1 SEA language.

For example, if the parallel MT dataloader has these languages: ind, sun, eng, spa, fil, mya, the dataloader consists of these subsets:
- `_{DATALOADERNAME}_eng_ind_seacrowd_t2t`
- `_{DATALOADERNAME}_eng_sun_seacrowd_t2t`
- `_{DATALOADERNAME}_eng_mya_seacrowd_t2t`
- `_{DATALOADERNAME}_ind_eng_seacrowd_t2t`
- `_{DATALOADERNAME}_sun_eng_seacrowd_t2t`
- `_{DATALOADERNAME}_mya_eng_seacrowd_t2t`
- `_{DATALOADERNAME}_spa_ind_seacrowd_t2t`
- `_{DATALOADERNAME}_spa_sun_seacrowd_t2t`
- `_{DATALOADERNAME}_spa_mya_seacrowd_t2t`
- `_{DATALOADERNAME}_ind_spa_seacrowd_t2t`
- `_{DATALOADERNAME}_sun_spa_seacrowd_t2t`
- `_{DATALOADERNAME}_mya_spa_seacrowd_t2t`
- `_{DATALOADERNAME}_ind_sun_seacrowd_t2t`
- `_{DATALOADERNAME}_ind_mya_seacrowd_t2t`
- `_{DATALOADERNAME}_sun_ind_seacrowd_t2t`
- `_{DATALOADERNAME}_mya_ind_seacrowd_t2t`
- `_{DATALOADERNAME}_sun_mya_seacrowd_t2t`
- `_{DATALOADERNAME}_mya_sun_seacrowd_t2t`

But since `eng` and `spa` are both non-SEA languages, no need to include `_{DATALOADERNAME}_eng_spa_seacrowd_t2t` and `_{DATALOADERNAME}_spa_eng_seacrowd_t2t`.

cc: @SamuelCahyawijaya @sabilmakbar

**Comment by Enliven26** [comment]
> > Ups @Enliven26, sorry for the last comment, I think we have a conflicting argument here. Let me discuss this internally first with @holylovenia and the others regarding which subsets should we take 🙏🏻
> 
> Hi @Enliven26, apologies for the previous contradictory advice. 🙏 After discussing, this is what we agreed on.
> 
> For parallel MT dataloaders, we agreed upon having a subset for every possible direction with at least 1 SEA language.
> 
> For example, if the parallel MT dataloader has these languages: ind, sun, eng, spa, fil, mya, the dataloader consists of these subsets:
> 
> * `_{DATALOADERNAME}_eng_ind_seacrowd_t2t`
> * `_{DATALOADERNAME}_eng_sun_seacrowd_t2t`
> * `_{DATALOADERNAME}_eng_mya_seacrowd_t2t`
> * `_{DATALOADERNAME}_ind_eng_seacrowd_t2t`
> * `_{DATALOADERNAME}_sun_eng_seacrowd_t2t`
> * `_{DATALOADERNAME}_mya_eng_seacrowd_t2t`
> * `_{DATALOADERNAME}_spa_ind_seacrowd_t2t`
> * `_{DATALOADERNAME}_spa_sun_seacrowd_t2t`
> * `_{DATALOADERNAME}_spa_mya_seacrowd_t2t`
> * `_{DATALOADERNAME}_ind_spa_seacrowd_t2t`
> * `_{DATALOADERNAME}_sun_spa_seacrowd_t2t`
> * `_{DATALOADERNAME}_mya_spa_seacrowd_t2t`
> * `_{DATALOADERNAME}_ind_sun_seacrowd_t2t`
> * `_{DATALOADERNAME}_ind_mya_seacrowd_t2t`
> * `_{DATALOADERNAME}_sun_ind_seacrowd_t2t`
> * `_{DATALOADERNAME}_mya_ind_seacrowd_t2t`
> * `_{DATALOADERNAME}_sun_mya_seacrowd_t2t`
> * `_{DATALOADERNAME}_mya_sun_seacrowd_t2t`
> 
> But since `eng` and `spa` are both non-SEA languages, no need to include `_{DATALOADERNAME}_eng_spa_seacrowd_t2t` and `_{DATALOADERNAME}_spa_eng_seacrowd_t2t`.
> 
> cc: @SamuelCahyawijaya @sabilmakbar

given `_{DATALOADERNAME}_eng_ind_seacrowd_t2t` as example, should I include the transcript language for the subset so that it becomes `_{DATALOADERNAME}_eng_Latn_ind_Arab_seacrowd_t2t` for example? If not, then what is the default transcript language for both text?

**Comment by holylovenia** [comment]
> given `_{DATALOADERNAME}_eng_ind_seacrowd_t2t` as example, should I include the transcript language for the subset so that it becomes `_{DATALOADERNAME}_eng_Latn_ind_Arab_seacrowd_t2t` for example? If not, then what is the default transcript language for both text?

Yes yes, please use the transcript as you described as well. 🙏 Thank you so much, @Enliven26!!

**Comment by Enliven26** [comment]
Hi, please review my recent commit. Thanks!

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+    SEACROWD_VERSION = datasets.Version(_SEACROWD_VERSION)
+
+    BUILDER_CONFIGS = []
+
+    for first_lang_name in _LANGUAGE_NAMES:
+        for second_lang_name in _LANGUAGE_NAMES:
+            if first_lang_name == second_lang_name or ((first_lang_name.split("_")[0] not in _LANGUAGES) and (second_lang_name.split("_")[0] not in _LANGUAGES)):
+                continue
+
+            subset_id = f"{_DATASETNAME}_{first_lang_name}_{second_lang_name}_seacrowd_t2t"
```
Comment (SamuelCahyawijaya) [inline seacrowd/sea_datasets/flores200/flores200.py:347]: The current subset name looks like: `flores200_*_seacrowd_t2t_seacrowd_t2t` and ` 'flores200_*_seacrowd_t2t_source`, we can remove the `_seacrowd_t2t` here to fix the problem

File: seacrowd/sea_datasets/flores200/flores200.py [hunk 8]
```diff
@@ -0,0 +1,478 @@
...
+            for split in _SPLITS
+        ]
+
+    def _generate_examples(self, sentence_paths: list[str], metadata_path: str) -> Tuple[int, Dict]:
+        """Yields examples as (key, example) tuples."""
+
+        sentences = {}
+        langs = [self.config.first_language_name, self.config.second_language_name]
+
+        for path, lang in zip(sentence_paths, langs):
```
Comment (SamuelCahyawijaya) [inline seacrowd/sea_datasets/flores200/flores200.py:451]: The `sentence_path` provides the paths to the data for all available languages in Flores, while the `langs` is only lang1 and lang2, this causes misalignment between the data and the language code as shown below:

zsm_Latn - ind_Latn
```
>>> x = datasets.load_dataset('seacrowd/sea_datasets/flores200/flores200.py', name='flores200_zsm_Latn_ind_Latn_seacrowd_t2t_seacrowd_t2t')
>>> x['dev'][0]
{'id': '1', 'text_1': 'يق أورو سنين، اوق علمون دري فکولتس کدوکترن يونيۏرسيتس ستانفورد ݢڤعموم اکن جتومى الت دياݢنوستيک بارو ڽڠ جوت ݢڤيليه اتو سيل منوروت جنيهجيه: چيڤ اوبيت ڽڠ جوت ݢچيتق ڽڠ جوت ݢڤݢت دڠون ڤرينتر ايڠ\u200cکجيت ستندر دڠون يوم موڠکين نا سابوه سين ا.س. لم تيڤ سابوه.', 'text_2': 'Bak uroe Senin, awak ilmuwan dari Fakultas Kedokteran Universitas Stanford geu peuumum akan jiteumèe alat diagnostik baro nyang jeuet geupileh atoe sel meunurot jeunèhjih: chip ubit nyang jeuet geucetak nyang jeuet geupeuget deungôn printer inkjèt standar deungôn yum mungken na sabôh sen AS lam tiep sabôh.', 'text_1_name': 'zsm_Latn', 'text_2_name': 'ind_Latn'}
```

eng_Latn - ind_Latn
```
>>> x = datasets.load_dataset('seacrowd/sea_datasets/flores200/flores200.py', name='flores200_eng_Latn_ind_Latn_seacrowd_t2t_seacrowd_t2t')
>>> x['dev'][0]
{'id': '1', 'text_1': 'يق أورو سنين، اوق علمون دري فکولتس کدوکترن يونيۏرسيتس ستانفورد ݢڤعموم اکن جتومى الت دياݢنوستيک بارو ڽڠ جوت ݢڤيليه اتو سيل منوروت جنيهجيه: چيڤ اوبيت ڽڠ جوت ݢچيتق ڽڠ جوت ݢڤݢت دڠون ڤرينتر ايڠ\u200cکجيت ستندر دڠون يوم موڠکين نا سابوه سين ا.س. لم تيڤ سابوه.', 'text_2': 'Bak uroe Senin, awak ilmuwan dari Fakultas Kedokteran Universitas Stanford geu peuumum akan jiteumèe alat diagnostik baro nyang jeuet geupileh atoe sel meunurot jeunèhjih: chip ubit nyang jeuet geucetak nyang jeuet geupeuget deungôn printer inkjèt standar deungôn yum mungken na sabôh sen AS lam tiep sabôh.', 'text_1_name': 'eng_Latn', 'text_2_name': 'ind_Latn'}
```

**Review by SamuelCahyawijaya — CHANGES_REQUESTED** [review]
Hi @Enliven26, thank you for updating the code!

It seems like there are two problems, one with the subset naming and the other one with the loaded dataset.

I have put some comments on the code, please update them accordingly and let me know if there is any problem.

**Comment by Enliven26** [comment]
Hi @SamuelCahyawijaya , sorry for the misaligment error. I already fixed it, please review the changes. Thanks!

**Review by holylovenia — APPROVED** [review]
LGTM! Thanks a lot, @Enliven26!

**Review by SamuelCahyawijaya — APPROVED** [review]
@Enliven26 : Thank you for the updates! LGTM!

**Pull request merged**